### PR TITLE
openstack: fix cognitive complexity

### PR DESCRIPTION
**Pull request by mmartinv**
Signed-off-by: Miguel Martín <mmartinv@redhat.com>

### CODE REVIEW

File: pkg/controller/plan/adapter/openstack/builder.go [hunk 1]
```diff
@@ -894,55 +895,61 @@ func (r *Builder) PopulatorVolumes(vmRef ref.Ref, annotations map[string]string,
...
 		if err != nil {
-			if !k8serr.IsNotFound(err) {
-				err = liberr.Wrap(err)
+			if !k8serr.IsAlreadyExists(err) {
+				err = liberr.Wrap(err, "couldn't build the PVC",
+					"image", image.Name)
 				return
 			}
+			err = nil
+			continue
```
Comment (ahadas) [inline pkg/controller/plan/adapter/openstack/builder.go:911]: shouldn't we proceed to `pvcNames = append(pvcNames, pvc.Name)` in this case?

File: pkg/controller/plan/adapter/openstack/builder.go [hunk 2]
```diff
@@ -894,55 +895,61 @@ func (r *Builder) PopulatorVolumes(vmRef ref.Ref, annotations map[string]string,
...
-		populatorName, err = r.createVolumePopulatorCR(image, secretName, vmRef.ID)
+		populatorName, err = r.createVolumePopulatorCR(*image, secretName, workload.ID)
 		if err != nil {
 			err = liberr.Wrap(err)
 			return
 		}
+		pvc, err = r.ensurePopulatorPVC(workload, image, annotations, populatorName, secretName)
+		return
+	}
+	err = k8serr.NewAlreadyExists(schema.ParseGroupResource("openstackvolumepopulator.forklift.konveyor.io"), populator.Name)
```
Comment (ahadas) [inline pkg/controller/plan/adapter/openstack/builder.go:934]: I'd expect 'ensure' method to return err == nil when the entity already exists - this behavior is surprising

Comment (mmartinv) [reply pkg/controller/plan/adapter/openstack/builder.go:934]: Yeah maybe that's better approach, but then we would need to check if the returned pvc is `nil` or not

Comment (mmartinv) [reply pkg/controller/plan/adapter/openstack/builder.go]: Please take look now, I think it's better now.

File: pkg/controller/plan/adapter/openstack/builder.go (hunk 1 above)
Comment (mmartinv) [reply pkg/controller/plan/adapter/openstack/builder.go]: outdated, please check the latest push

**Review by ahadas — COMMENTED** [review]
commented but I'll address those comments

File: pkg/controller/plan/adapter/openstack/client.go [hunk 3]
```diff
@@ -932,3 +726,242 @@ func (r *Client) getImagesFromVolumes(vm *libclient.VM) (images []libclient.Imag
...
+func (r *Client) ensureImagesFromVolumesReady(vm *libclient.VM) (ready bool, err error) {
+	var imagesFromVolumes []libclient.Image
+	imagesFromVolumes, err = r.getImagesFromVolumes(vm)
+	if err != nil {
+		err = liberr.Wrap(err)
+		r.Log.Error(err, "error while trying to get the images from the VM volumes",
+			"vm", vm.Name)
+		return
+	}
+	ready = true
```
[the remark this answers is not in the document]
Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/client.go]: ok, that was an incorrect observation but there was still room for improvement here

**Comment by ahadas** [comment]
@bennyz please review the openstack related changes
@liranr23 please review the changes in EnsurePopulatorVolumes (filtering pending PVCs for the binding pod)

File: pkg/controller/plan/adapter/openstack/builder.go [hunk 4]
```diff
@@ -893,56 +891,64 @@ func (r *Builder) PopulatorVolumes(vmRef ref.Ref, annotations map[string]string,
...
-		if err != nil {
-			if !k8serr.IsNotFound(err) {
-				err = liberr.Wrap(err)
-				return
-			}
-		}
+func (r *Builder) getCorrespondingPvc(image model.Image, workload *model.Workload, vmRef ref.Ref, annotations map[string]string, secretName string) (*core.PersistentVolumeClaim, error) {
+	if populatorName, err := r.ensureVolumePopulator(workload, &image, secretName); err == nil {
+		return r.ensureVolumePopulatorPVC(workload, &image, annotations, populatorName)
+	} else {
```
Comment (bennyz) [inline pkg/controller/plan/adapter/openstack/builder.go:912]: if used named return value we can dropped the else here

Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/builder.go:912]: but with named return values, I'll need to define populatorName separately... I can drop the 'else' now as well..

Comment (bennyz) [reply pkg/controller/plan/adapter/openstack/builder.go:912]: but err will be out of scope (unless defined at the top), no?

Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/builder.go:912]: ah right right

Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/builder.go]: changed, also resolved a conflict with #591 - please double check me

**Comment by liranr23** [comment]
> @liranr23 please review the changes in EnsurePopulatorVolumes (filtering pending PVCs for the binding pod)

looks ok, you can also check the storage class (or profile) and see if it's wait for consumer.

**Review by liranr23 — APPROVED** [review]

**Review by bennyz — APPROVED** [review]

**Pull request merged**